asllib/ASTUtils.ml

```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
     [] li1
   |> List.rev
 
+let list_flat_cross f li1 li2 =
+  List.fold_left
+    (fun xys x ->
+      List.fold_left (fun xys' y -> List.rev_append (f x y) xys') xys li2)
+    [] li1
+  |> List.rev
+
 let is_left_increasing = function
-  | MUL | DIV | DIVRM | SHL | SHR | POW | PLUS | MINUS -> true
-  | AND | BAND | BEQ | MOD | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MUL | SHL | SHR | POW | PLUS | MINUS -> true
+  | DIV | DIVRM | AND | BAND | BEQ | MOD | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_increasing = function
   | MUL | SHL | SHR | POW | PLUS -> true
-  | DIV | DIVRM | MINUS -> false
-  | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MINUS -> false
+  | DIV | DIVRM | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_decreasing = function
```

Comment:
Do we even need this function? It's the boolean negation of is_right_incvreasing.